train.py

```diff
@@ -18,22 +20,51 @@
 
 sess.run(tf.global_variables_initializer())
 
+
 # Training loop variables
 epochs = 100
 batch_size = 50
 num_samples = data.num_examples
 step_size = int(num_samples / batch_size)
 
+
+early_stop_steps = 200
+best_score       = np.inf
+val_loss_value   = np.inf
+best_idx         = 0
+stop_flag        = False
+
 for epoch in range(epochs):
+    if stop_flag:
+        break
+
     for i in range(step_size):
-        batch = data.next_batch(100)
+        batch = data.next_batch(batch_size)
+        cur_step = epoch * step_size + i
+
 
         train_step.run(feed_dict={model.x: batch[0], model.y_: batch[1], model.keep_prob: 0.8})
 
+        ## Update early stopping params
+        if val_loss_value <= best_score:
+          best_idx   = epoch * step_size + i
+          best_score = val_loss_value
+
+
         if i%10 == 0:
-          loss_value = loss.eval(feed_dict={model.x:batch[0], model.y_: batch[1], model.keep_prob: 1.0})
-          print("epoch: %d step: %d loss: %g"%(epoch, epoch * batch_size + i, loss_value))
+          train_loss_value = loss.eval(feed_dict={model.x:batch[0], model.y_: batch[1], model.keep_prob: 1.0})
+          val_loss_value   = loss.eval(feed_dict={model.x:val[0], model.y_:val[1], model.keep_prob: 1.0})
+          print("epoch: %d step: %d train_loss: %g val_loss: %g best_idx: %g best_score: %g"%
+            (epoch, cur_step, train_loss_value, val_loss_value, best_idx, best_score))
+
+        ## Exit loop if early_stopping criteria is met
+        if (cur_step - best_idx) >= early_stop_steps:
+          stop_flag = True
+          break
```

Comment:
I am not sure. I read this:
https://www.quora.com/How-does-one-employ-early-stopping-in-TensorFlow

Since it was already set up so that the training loop was written in python, I figured this was the easiest way. I got the impression that the tensorflow library does include methods to auto-iterate, etc but this only works for simple, pre-designed networks. That may be incorrect.